File: testdata/fixtures/kotlin/CWE-352.kt
// CWE-352: CSRF — Post mappings without token
package fixture
import org.springframework.web.bind.annotation.*

@RestController
class CsrfController {

    // ruleid: kotlin-csrf
    @PostMapping("/transfer")
    fun transfer(amount: String): String {
        return "transferred $amount"
    }

    // ruleid: kotlin-csrf
    @PostMapping("/update")
    fun update(email: String): String {
        return "updated $email"
    }

    // ruleid: kotlin-csrf
    @DeleteMapping("/delete")
    fun delete(id: String): String {
        return "deleted $id"
    }

    // ok: kotlin-csrf
    @PostMapping("/safe")
    fun safeTransfer(token: String, amount: String): String {
        verifyCsrf(token)
        return "ok"
    }

    // ok: kotlin-csrf
    @PostMapping("/safe2")
    fun safe2(csrfToken: String): String {
        val t = CsrfToken(csrfToken)
        return t.value
    }
}

File: testdata/fixtures/ocaml/CWE-352.ml
(* CWE-352: CSRF - Dream.post without token *)
let vuln1 =
  (* ruleid: ocaml-csrf *)
  Dream.post "/transfer" (fun req -> Dream.html "transferred")

let vuln2 =
  (* ruleid: ocaml-csrf *)
  Dream.post "/update" (fun req -> Dream.html "updated")

let vuln3 =
  (* ruleid: ocaml-csrf *)
  Dream.post "/delete" (fun _ -> Dream.html "deleted")

let safe1 =
  (* ok: ocaml-csrf *)
  Dream.post "/safe" (fun req ->
    match Dream.header req "X-CSRF-Token" with
    | Some t when t = Dream.csrf_token req -> Dream.html "ok"
    | _ -> Dream.html "forbidden")

let safe2 =
  (* ok: ocaml-csrf *)
  Dream.post "/safe2" (fun req ->
    let token = Dream.csrf_token req in
    Dream.html token)
